confdb/confdb.go

```diff
@@ -1387,62 +1559,148 @@ func (p viewRule) isWriteable() bool {
-	String() string
+	accessor
+
+	match(subkey string, matched *matchedPlaceholders) bool
+}
+
+type writeOpts struct {
+	topLevel bool
 }
 
 type storageWriter interface {
-	write(sb *strings.Builder, placeholders map[string]string) error
+	accessor
+
+	write(sb *strings.Builder, matched *matchedPlaceholders, opts writeOpts)
 }
 
 // placeholder represents a subkey of a name/path (e.g., "{foo}") that can match
 // with any value and map it from the input name to the path.
-type placeholder string
+type keyPlaceholder string
 
-// match adds a mapping to the placeholders map from this placeholder key to the
+// match adds an entry to matchedPlaceholders mapping this placeholder key to the
 // supplied name subkey and returns true (a placeholder matches with any value).
-func (p placeholder) match(subkey string, placeholders map[string]string) bool {
-	placeholders[string(p)] = subkey
+func (p keyPlaceholder) match(subkey string, matched *matchedPlaceholders) bool {
+	if matched.key == nil {
+		matched.key = make(map[string]string)
+	}
+	matched.key[string(p)] = subkey
 	return true
 }
 
-// write writes the value from the placeholders map corresponding to this placeholder
-// key into the strings.Builder.
-func (p placeholder) write(sb *strings.Builder, placeholders map[string]string) error {
-	subkey, ok := placeholders[string(p)]
+// write writes the value from the matchedPlaceholders entry corresponding to
+// this placeholder key into the strings.Builder.
+func (p keyPlaceholder) write(sb *strings.Builder, matched *matchedPlaceholders, opts writeOpts) {
+	subkey, ok := matched.key[string(p)]
 	if !ok {
 		// placeholder wasn't matched, return the original key in brackets
-		subkey = fmt.Sprintf("{%s}", string(p))
+		subkey = p.access()
 	}
 
-	_, err := sb.WriteString(subkey)
-	return err
+	if !opts.topLevel {
+		sb.WriteRune('.')
+	}
+
+	sb.WriteString(subkey)
 }
 
-// String returns the placeholder as a string.
-func (p placeholder) String() string {
+func (p keyPlaceholder) access() string {
 	return "{" + string(p) + "}"
 }
 
-// literal is a non-placeholder name/path subkey.
-type literal string
+func (p keyPlaceholder) name() string     { return string(p) }
+func (p keyPlaceholder) keyType() keyType { return keyPlaceholderType }
+
+type matchedPlaceholders struct {
+	index map[string]string
+	key   map[string]string
+}
+
+// indexPlaceholder represents a subkey of a name/path (e.g., "[{n}]") that can
+// match an index value and map it from the input name to the path.
+type indexPlaceholder string
+
+// match checks if the subkey can be used to index a list. If so, it adds an
+// entry to matchedPlaceholders mapping this placeholder key to the supplied
+// name subkey and returns true.
+func (p indexPlaceholder) match(subkey string, matched *matchedPlaceholders) bool {
+	subkey, ok := stripIndex(subkey)
+	if !ok {
+		return false
+	}
+
+	if _, err := strconv.Atoi(subkey); err != nil {
```

Comment:
shouldn't index placeholders be ascii e.g. `[{n}]`? I might be confusing something else.

Reply:
also seems to be missing a unit test for errors

Reply:
This is the matching of a path part to that index placeholder, so we expect the provided piece that matches to that placeholder to be an int. This is going to be dropped in the next PR. Basically I'll use the new types that implement `accessor` to simplify the matching and drop the string parsing we do here so that will address the coverage issue (or if it doesn't I'd rather address it there if that's ok)